pkg/statistics/region_collection.go

```diff
@@ -42,84 +48,86 @@ const (
-	regionLearnerPeerRegionCounter    = regionStatusGauge.WithLabelValues("learner-peer-region-count")
-	regionEmptyRegionCounter          = regionStatusGauge.WithLabelValues("empty-region-count")
-	regionOversizedRegionCounter      = regionStatusGauge.WithLabelValues("oversized-region-count")
-	regionUndersizedRegionCounter     = regionStatusGauge.WithLabelValues("undersized-region-count")
-	regionWitnesssLeaderRegionCounter = regionStatusGauge.WithLabelValues("witness-leader-region-count")
-
-	offlineMissPeerRegionCounter    = offlineRegionStatusGauge.WithLabelValues("miss-peer-region-count")
-	offlineExtraPeerRegionCounter   = offlineRegionStatusGauge.WithLabelValues("extra-peer-region-count")
-	offlineDownPeerRegionCounter    = offlineRegionStatusGauge.WithLabelValues("down-peer-region-count")
-	offlinePendingPeerRegionCounter = offlineRegionStatusGauge.WithLabelValues("pending-peer-region-count")
-	offlineLearnerPeerRegionCounter = offlineRegionStatusGauge.WithLabelValues("learner-peer-region-count")
-	offlineOfflinePeerRegionCounter = offlineRegionStatusGauge.WithLabelValues("offline-peer-region-count")
+	regionMissPeerRegionCounter      = regionStatusGauge.WithLabelValues("miss-peer-region-count")
+	regionExtraPeerRegionCounter     = regionStatusGauge.WithLabelValues("extra-peer-region-count")
+	regionDownPeerRegionCounter      = regionStatusGauge.WithLabelValues("down-peer-region-count")
+	regionPendingPeerRegionCounter   = regionStatusGauge.WithLabelValues("pending-peer-region-count")
+	regionLearnerPeerRegionCounter   = regionStatusGauge.WithLabelValues("learner-peer-region-count")
+	regionEmptyRegionCounter         = regionStatusGauge.WithLabelValues("empty-region-count")
+	regionOversizedRegionCounter     = regionStatusGauge.WithLabelValues("oversized-region-count")
+	regionUndersizedRegionCounter    = regionStatusGauge.WithLabelValues("undersized-region-count")
+	regionWitnessLeaderRegionCounter = regionStatusGauge.WithLabelValues("witness-leader-region-count")
+	regionOfflinePeerRegionCounter   = offlineRegionStatusGauge.WithLabelValues("offline-peer-region-count")
 )
 
-// RegionInfo is used to record the status of region.
-type RegionInfo struct {
-	*core.RegionInfo
+// RegionInfoWithTS is used to record the extra timestamp status of a region.
+type RegionInfoWithTS struct {
+	id                   uint64
 	startMissVoterPeerTS int64
 	startDownPeerTS      int64
 }
 
 // RegionStatistics is used to record the status of regions.
 type RegionStatistics struct {
 	sync.RWMutex
-	conf               sc.CheckerConfig
-	stats              map[RegionStatisticType]map[uint64]*RegionInfo
-	offlineStats       map[RegionStatisticType]map[uint64]*core.RegionInfo
+	rip   RegionInfoProvider
+	conf  sc.CheckerConfig
+	stats map[RegionStatisticType]map[uint64]*RegionInfoWithTS
+	// Since we may easily have a large number of offline regions in the scale-in scenario,
```

Comment:
Will the empty region have the same problem?

Reply:
All abnormal region statuses will have the same problem, but since the offline peer is the most common case and easily triggered, I only handled it here.

Reply:
If so, maybe we can handle the offline peer in the same way.

Reply:
Make sense, I have addressed it, PTAL again.